Documentation/guides/cpp_cmake.rst

```diff
@@ -82,159 +87,63 @@ Creating the project
-    set(BUILD_SHARED_LIBS OFF)
-    add_subdirectory(src)
-
-* hellocpp/cmake/stm32f4discovery.cmake
-
-.. code-block:: cmake
-    
-    set(CMAKE_SYSTEM_NAME Generic)
-    set(CMAKE_SYSTEM_PROCESSOR arm)
-
-    set(MCU_LINKER_SCRIPT "${NUTTX_PATH}/scripts/ld.script")
-
-    set(COMPILER_PREFIX arm-none-eabi-)
-
-    # cmake-format: off
-    set(CMAKE_C_COMPILER    ${COMPILER_PREFIX}gcc)
-    set(CMAKE_CXX_COMPILER  ${COMPILER_PREFIX}g++)
-    set(CMAKE_AR            ${COMPILER_PREFIX}ar)
-    set(CMAKE_RANLIB        ${COMPILER_PREFIX}ranlib)
-    set(CMAKE_LINKER        ${COMPILER_PREFIX}ld)
-    set(CMAKE_ASM_COMPILER  ${COMPILER_PREFIX}gcc)
-    set(CMAKE_OBJCOPY       ${COMPILER_PREFIX}objcopy)
-    set(CMAKE_OBJDUMP       ${COMPILER_PREFIX}objdump)
-    set(CMAKE_SIZE          ${COMPILER_PREFIX}size)
-
-    set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
-    set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
-    set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
-    set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
-
-    set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
-
-    set(AC_HW_FLAGS         "-mcpu=cortex-m4 -mthumb -mfloat-abi=soft ")
-    set(AC_HW_FLAGS         "${AC_HW_FLAGS} -isystem ${NUTTX_PATH}/include")
-    set(AC_HW_FLAGS         "${AC_HW_FLAGS} -pipe")
-
-    set(AC_LINKER_FLAGS     "--entry=__start -nostdlib -T${MCU_LINKER_SCRIPT}")
-
-* hellocpp/src/CMakeLists.txt
-
-.. code-block:: cmake
-   
-    set(HEADER_FILES
-            HelloWorld.h
-    )
 
     set(SOURCE_FILES
-            HelloWorld.cpp
+            ${CMAKE_CURRENT_SOURCE_DIR}/HelloWorld.cpp
+            ${CMAKE_CURRENT_SOURCE_DIR}/main.cpp
     )
 
-    link_directories(${EXE_NAME} ${NUTTX_PATH}/libs)
-
-    add_executable(${EXE_NAME} ${SOURCE_FILES} main.cpp ${HEADER_FILES})
+    set(EXE_NAME "hello")
+    add_executable(${EXE_NAME} ${SOURCE_FILES})
 
     add_custom_command(
             TARGET ${EXE_NAME}
             POST_BUILD
             COMMAND ${CMAKE_OBJCOPY} ARGS -S -O binary ${CMAKE_BINARY_DIR}/${EXE_NAME}.elf ${CMAKE_BINARY_DIR}/${EXE_NAME}.bin
-    )
-
-    target_link_libraries(${EXE_NAME} --start-group)
-
-    target_link_libraries(${EXE_NAME} sched)
-    target_link_libraries(${EXE_NAME} drivers)
-    target_link_libraries(${EXE_NAME} boards)
-    target_link_libraries(${EXE_NAME} c)
-    target_link_libraries(${EXE_NAME} mm)
-    target_link_libraries(${EXE_NAME} arch)
-    target_link_libraries(${EXE_NAME} xx)
-    target_link_libraries(${EXE_NAME} apps)
-    target_link_libraries(${EXE_NAME} fs)
-    target_link_libraries(${EXE_NAME} binfmt)
-    target_link_libraries(${EXE_NAME} board)
-    target_link_libraries(${EXE_NAME} gcc)
-    target_link_libraries(${EXE_NAME} supc++)
 
-    target_link_libraries(${EXE_NAME} --end-group)
 
-* hellocpp/src/main.cpp
+* hellocpp/main.cpp
 
 .. code-block:: c++
 
+    #include <memory>
+
     #include "HelloWorld.h"
-    #include <nuttx/config.h>
 
-    extern "C"
+    extern "C" int main(int, char*[])
```

Comment:
`extern C` is not necessary, could we remove it?

Reply:
it need since main will expand to xxx_main, which make c++ compiler can't skip the name mangle.